File: LeetCode/26_remove_duplicates.rb
require 'set'

def three_sum(nums)
    missing = Hash.new
    sums = Set[]
    nums.each {|num| missing[0-num] = true}
    (0...nums.length).each do |i|
        (0...i).each do |j|
            if missing[nums[i] + nums[j]]
                (0...j).each do |k|
                    these_nums =[nums[i], nums[j], nums[k]]
                    sums.add(these_nums.sort) if these_nums.sum == 0
                end
            end
        end
    end
    sums
end

p three_sum([-1,0,1,2,-1,-4])

File: LeetCode/26_remove_duplicates.py
def three_sum(nums):
    missing = {}
    sums = []
    for num in nums:
        missing[0-num] = True
    for i in range(len(nums)):
        for j in range(i):
            if (nums[i] + nums[j]) in missing:
                for k in range(j):
                    these_nums = [nums[i],nums[j],nums[k]]
                    these_nums.sort()
                    if sum(these_nums) == 0:
                        if these_nums not in sums: sums.append(these_nums)
    return sums

print(three_sum([-1,0,1,2,-1,-4]))